File: Jacobi.m
function [E, s] = Jacobi(x0, A, b, Tol, niter, tipoError)
    E = [];
    s = [];

    try
        % Validaciones
        if nargin < 6
            error('Faltan argumentos. Debes ingresar: x0, A, b, Tol, niter, tipoError');
        end

        if ~ismatrix(A) || size(A,1) ~= size(A,2)
            error('La matriz A debe ser cuadrada.');
        end

        if length(b) ~= size(A,1) || length(x0) ~= size(A,1)
            error('Las dimensiones de x0, A y b no son compatibles.');
        end

        if Tol <= 0
            error('La tolerancia debe ser un número positivo.');
        end

        if niter <= 0 || floor(niter) ~= niter
            error('niter debe ser un entero positivo.');
        end

        if ~strcmpi(tipoError, 'relativo') && ~strcmpi(tipoError, 'absoluto')
            error('tipoError debe ser "relativo" o "absoluto".');
        end

        % Verificar diagonal dominante
        isDD = true;
        for i = 1:size(A,1)
            diagElem = abs(A(i,i));
            offDiagSum = sum(abs(A(i,:))) - diagElem;
            if diagElem <= offDiagSum
                isDD = false;
                break;
            end
        end
        if ~isDD
            warning('La matriz A no es diagonalmente dominante. El método podría no converger.');
        end

        % Inicialización
        D = diag(diag(A));
        L = -tril(A, -1);
        U = -triu(A, 1);
        T = inv(D) * (L + U);
        C = inv(D) * b;
        c = 0;
        error_val = Tol + 1;
        n = length(x0);

        X = zeros(n, niter + 1); % Matriz para guardar evolución de las variables
        X(:, 1) = x0;

        % Encabezado de tabla
        fprintf('--------------------------------------------------------------------------------------------------\n');
        fprintf('| Iter |');
        for i = 1:n
            fprintf('      x(%d)       |', i);
        end
        fprintf('     Error (%s)     |\n', tipoError);
        fprintf('--------------------------------------------------------------------------------------------------\n');

        % Primera fila (x0)
        fprintf('| %4d |', c);
        for i = 1:n
            fprintf(' %14.6f |', x0(i));
        end
        fprintf('         ---         |\n');

        % Iteraciones
        while error_val > Tol && c < niter
            x1 = T * x0 + C;

            if strcmpi(tipoError, 'relativo')
                error_val = norm(x1 - x0, inf) / max(norm(x1, inf), eps);
            else
                error_val = norm(x1 - x0, inf);
            end

            E(c + 1) = error_val;
            c = c + 1;
            x0 = x1;
            X(:, c + 1) = x0; % almacenar siguiente aproximación

            fprintf('| %4d |', c);
            for i = 1:n
                fprintf(' %14.6f |', x0(i));
            end
            fprintf(' %18.10e |\n', E(c));
        end

        fprintf('--------------------------------------------------------------------------------------------------\n');

        % Salida
        s = x0;
        if error_val < Tol
            fprintf('\nJacobi: El método convergió. Solución aproximada con tolerancia %.2e:\n', Tol);
            disp(s');
        else
            fprintf('\nJacobi: El método fracasó después de %d iteraciones.\n', niter);
        end

        % Gráfica de evolución de variables
        figure;
        plot(0:c, X(:, 1:c+1)', '-o', 'LineWidth', 1.5);
        xlabel('Iteración');
        ylabel('Valor de las variables');
        legend(arrayfun(@(i) sprintf('x(%d)', i), 1:n, 'UniformOutput', false));
        title('Evolución de las variables (Jacobi)');
        grid on;

    catch ME
        fprintf('Error en Jacobi: %s\n', ME.message);
    end
end

File: analisis_numerico/metodos/algoritmos/jacobi.py
import numpy as np
import io
import base64
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

def jacobi(x0, A, b, Tol, niter, error_type):
    # Paso 1
    A = np.array(A, float)
    b = np.array(b, float)
    x = np.array(x0, float)

    if not (A.ndim == 2 and A.shape[0] == A.shape[1]):
        raise ValueError("La matriz A debe ser cuadrada.")
    if b.size != A.shape[0] or x.size != A.shape[0]:
        raise ValueError("Las dimensiones de x0, A y b no son compatibles.")
    if error_type.lower() not in ('absoluto', 'relativo'):
        raise ValueError("error_type debe ser 'absoluto' o 'relativo'.")

    historial = []
    n = A.shape[0]
    D = np.diag(A)
    R = A - np.diagflat(D)
    X = np.zeros((n, int(niter)+1))
    X[:, 0] = x

    for k in range(1, int(niter)+1):
        x_new = (b - R.dot(x)) / D
        if error_type.lower() == 'absoluto':
            E = np.linalg.norm(x_new - x, ord=np.inf)
        else:
            E = np.linalg.norm((x_new - x) / x_new, ord=np.inf)
        historial.append((k, x_new.copy(), E))
        X[:, k] = x_new
        x = x_new
        if E < Tol:
            break

    # Gráfica
    plt.figure()
    for i in range(n):
        plt.plot(range(0, k+1), X[i, :k+1], '-o', linewidth=1.5, label=f'x({i+1})')
    plt.xlabel('Iteración')
    plt.ylabel('Valor de las variables')
    plt.legend(loc='best')
    plt.title('Evolución de las variables (Jacobi)')
    plt.grid(True)

    buf = io.BytesIO()
    plt.savefig(buf, format='png')
    plt.close()
    buf.seek(0)
    img_base64 = base64.b64encode(buf.read()).decode('ascii')

    return {
        'x': x.tolist(),
        'historial': historial,
        'grafica_base64': img_base64
    }